### ISSUE TITLE
Name Collisions

### ISSUE THREAD

**Issue by pburgmer** [vojtechhabarta/typescript-generator#99]
Hi,

First of all: thanks for this great plugin and all your work.

We use a lot of inner classes. Of course they have a unique full qualified name (package, parent class and inner class name). But we use the same names multiple times within different parent classes. This causes ```Multiple classes are mapped to 'XYZ' name.``` error messages.

Is it possible to configure the plugin to respect the FQN and use it in the TypeScript output. E.g. as nested namespaces or multiple ambient module declarations?

Best regards,
Philipp

**Comment by vojtechhabarta** [comment]
Hi Philipp, thanks for your interest in this tool.

I think this feature would complicate typescript-generator internally and also for users. It already supports number of module/namespace variants (see [Modules and Namespaces](http://www.habarta.cz/typescript-generator/doc/ModulesAndNamespaces.html) Wiki page) and "complete" support for this feature would add next dimension to this matrix. And I also had request for dividing output to files by classes which is another dimension to this problem.

But I understand your motivation so lets leave this issue open and see what are other people's use cases and how all this could be implemented in single and simple way.

First question is: should Java packages be mapped to namespaces or modules?
Another one is how Java packages should be used. They usually form deep hierarchies and using whole package isn't probably desired. So some customization will be needed.

What do you think?
And does anybody have some use case or proposal?

**Comment by pburgmer** [comment]
Hi,

thank you for the quick response.

I saw the module/namespace variants. You're right, adding more options will not make the tool easier to use. Unfortunately not having this feature, the tool feels incomplete to me. I was really surprised to see the tool does not handle packages in the output.

How do you avoid name collisions? How do you ensure unique Class-names in your Java code? By convention? Or let the build crash and then refactor? In my opinion let the java compiler handle this issue is the easiest way. Perhaps I have overlooked something?

Mapping to namespaces or modules shouldn't make a big difference. As far as I know both can be defined pseudo nested. Each for itself, all below each other.

For namespaces we can use same dot-notation used by Java package names. See GreetingLib.Options at [organizing-types](https://www.typescriptlang.org/docs/handbook/declaration-files/by-example.html#organizing-types) in the declaration files documentation.

For Modules I would prefer the ES6 module notation with slashes. I tested a module declaration with `declare module "com/project/server" {}` and was able to import interfaces from there. So each package could be a separate module declaration, all top level in the d.ts file with the full package name as module name and dots converted to slashes. Like Java handles the packages at file system level ;)

Please tell me if I can assist you with implementing this feature. Maybe I can implement it and create a pull request if you prefer that. Or is it more an issue of adding this feature in general and not the implementation part?

**Comment by vojtechhabarta** [comment]
Hi,

First some current possibilities:
- If you have few random name conflicts you can use `customTypeNaming` to specify new names for conflicting classes. It is not needed to rename Java class.
- One use case where many conflicts might occur is when there are multiple versions of the same REST API. In this case you can invoke typescript-generator multiple times (each time for one version) and produce separate declaration file for each version.

Mapping Java packages to TypeScript namespaces seems natural apart from different casing conventions but this is not a big problem. For mapping to modules I see two options: slashes or dots.

What about nested classes? Imagine for example class `com.project.server.Person$Address`. Do you think it is better to map outer class to namespace/module like this:

``` typescript
declare namespace com.project.server.Person {
    class Address {}
}
```

or to create some joined name like `PersonAddress` or `Person$Address` like this:

``` typescript
declare namespace com.project.server {
    class Person$Address {}
}
```

If there will be some consensus how this feature should work I would like to implement it myself because there will be changes in core model (qualified identifiers will be needed throughout the tool).

**Comment by vojtechhabarta** [comment]
@pburgmer just wanted to note that I merged #113 (Custom type names specified using JavaScript function). It's still not what you wanted but it could help in some cases.

**Comment by pburgmer** [comment]
@vojtechhabarta 
Sorry for not responding. Too busy at the moment.
Feature sounds promising. Will give it a try asap.

**Comment by ollwenjones** [comment]
I had to use `customTypeNaming` to get around some name conflicts also, but it was a manageable number.

**Comment by vojtechhabarta** [comment]
I am finishing implementation of this feature and I just wanted to mention one constraint: In general it is not possible to emit each namespace all at once.

When generating TypeScript classes (not interfaces) each base class must be emitted before derived class.

Consider this case:
``` java
package com.p1;
class A {}
class C extends B {}
class D extends C {}

package com.p2;
class B extends A {}
```

In this case classes must be emitted in this order: `A`, `B`, `C`, `D`.

So I will not try to group all declarations from one namespace together. Instead I will emit namespace declaration for each class and interface even if it isn't absolutely needed. It is not so beautiful but it doesn't have impact on resulting structure.

So in this example case following output will be generated:
``` typescript
namespace com.p1 {
    class A {}
}
namespace com.p2 {
    class B extends com.p1.A {}
}
namespace com.p1 {
    class C extends com.p2.B {}
}
namespace com.p1 {
    class D extends com.p1.C {}
}
```

**Comment by pburgmer** [comment]
Yeah, sound promising! Thanks a lot.

**Comment by vojtechhabarta** [comment]
Released [1.24.314](https://github.com/vojtechhabarta/typescript-generator/releases/tag/v1.24.314).
